### ISSUE TITLE
cabal-cache exits with "thread blocked indefinitely"

### ISSUE THREAD

**Issue by hasufell** [haskell-works/cabal-cache#76]
```
cabal-cache: Network.Socket.recvBuf: resource vanished (Connection reset by peer)
cabal-cache: thread blocked indefinitely in an STM transaction
```

This happens quite regularly.

**Comment by newhoggy** [comment]
Is this still an issue?

**Comment by hasufell** [comment]
This happened with a poor internet connection in Asia. I'm not sure how to test it anymore. Maybe in a docker container?

**Comment by hasufell** [comment]
Still happens to me all over the place.

E.g. here: https://github.com/haskell/ghcup-hs/pull/717/checks?check_run_id=10199578543

**Comment by hasufell** [comment]
Maybe related:

* https://github.com/simonmar/async/issues/14
* https://gitlab.haskell.org/ghc/ghc/-/issues/9401

**Comment by hasufell** [comment]
Maybe it's this:

https://github.com/haskell-works/cabal-cache/blob/4ef410e1a095e1a08e3b801bc55b6948c5c8af53/src/HaskellWorks/CabalCache/Concurrent/Fork.hs#L8-L17

It looks like it would retry indefinitely if a thread crashed and never wrote `tDone`. The RTS will probably detect this as a deadlock then.

**Comment by newhoggy** [comment]
When we start a download we mark in our queue that a packaging is in the process of downloading and when it is complete, the packaged is marked as downloaded.

If an exception happens we fail do mark the download as failed.  The thread that was downloading has died so the download will never complete and other threads will wait indefinitely.

This PR will catch all exceptions and mark the package as failed to download.

https://github.com/haskell-works/cabal-cache/pull/192

**Comment by hasufell** [comment]
> When we start a download we mark in our queue that a packaging is in the process of downloading and when it is complete, the packaged is marked as downloaded.
> 
> If an exception happens we fail do mark the download as failed. The thread that was downloading has died so the download will never complete and other threads will wait indefinitely.
> 
> This PR will catch all exceptions and mark the package as failed to download.
> 
> #192

Maybe... but I still think using `finally` is the right thing to do in `forkThreadsWait`, because we shouldn't assume that the action never crashes.

**Comment by newhoggy** [comment]
Yeah, that makes sense.  Do you want a release with both changes in?

**Comment by hasufell** [comment]
> Yeah, that makes sense. Do you want a release with both changes in?

Yeah. I have a couple of other PRs upcoming though.

**Comment by newhoggy** [comment]
Are you intending to test these PRs with your own builds on `cabal-cache`?

**Comment by hasufell** [comment]
> Are you intending to test these PRs with your own builds of `cabal-cache`?

Yeah

**Comment by hasufell** [comment]
I did some testing with all 5 PRs applied on my main branch: https://github.com/hasufell/cabal-cache/tree/main

This works ok for me. I haven't given it a test in my CI yet.

**Comment by newhoggy** [comment]
Ok cool.  Let me know when you're ready for he PRs to merge and a new release.

**Comment by hasufell** [comment]
The retrying looks excessive: https://github.com/haskell/haskell-language-server/actions/runs/3740758812

cabal-cache has been killed on the darwin runner... The thing is, lots of 404 are legit, because there are no hits and it will spam retries. But which HTTP error codes make sense to retry on, I'm not sure. Maybe alternatively, we just disable retry by default.

**Comment by newhoggy** [comment]
Retries should exclude 404s if it doesn't already.

**Comment by newhoggy** [comment]
Of the `4xx` error codes the only one that seems like it should retry should be [429 Too Many Requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429)

**Comment by hasufell** [comment]
I propose to use a whitelist then:

```
  retryUnless (\appe -> case appErrorStatus appe of
                Just 402 -> False
                Just 408 -> False
                Just 410 -> False
                Just 425 -> False
                Just 429 -> False
                Just i
                  | i >= 500
                  , i < 600 -> False
                Just _   -> True
                Nothing  -> True) retries a
```

retry all 5xx and retry only 402, 408, 410, 425, 429 for 4xx.

**Comment by hasufell** [comment]
I think it works fine now: https://github.com/haskell/haskell-language-server/actions/runs/3746100155/jobs/6361150038

There was a minor logging bug left in the  retry PR (it would print it retries when it didn't), but hat's also fixed now. I'm suggesting to start the merge train.

**Comment by newhoggy** [comment]
In the end I went with the original solution, which is to catch all exceptions and mark the download failed.

**Comment by newhoggy** [comment]
Fixed in `1.0.5.4`.  Thanks for all your help with this issue!
